follow up on duplex operation for HTTP/2

wfurt opened the pull request:
This is follow-up on #36884 to address concerns about UnobservedTaskExceptions and error propagation. 

This change wraps all ignored tasks to make sure it does not end up in failed state without seeing the exception.   

I also added test case for case when exception happens while sending RequestBody after Response headers are received. When we hit plain IO exception, reading response will fail as it always did so I did not worry about is new test case.  

fixes #36855

wfurt committed:
follow up on duplex operation

stephentoub commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -297,6 +293,11 @@ public async Task WaitForConnectionShutdownAsync()
             _ignoreSettingsAck = false;
             _ignoreWindowUpdates = false;
+
+            if (frame != null)
```
Presumably you're moving this down so that we do all of the above cleanup even if frame is null.  Is it possible `await ReadFrameAsync` might throw, such that we should put the above in a finally block?  Or ReadFrameAsync won't return faulted tasks?

wfurt replied:
yes, that was the general idea. Closing socket kicks client side even if something goes wrong and test would fail instead of hanging. I can take another look at this.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1402,10 +1455,7 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
-                                Http2Connection c = (Http2Connection)state;
-                                if (NetEventSource.IsEnabled) c.Trace($"SendRequestBody Task failed. {t.Exception}");
-                             }, this, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
+                        _ = LogExceptionsAsync(bodyTask);
```
Do we need to make a similar call in the catch block above for the responseHeadersTask?  Otherwise, what happens if both tasks fault: we'll propagate the exception from the request task and will end up ignoring the response task.

wfurt replied:
done.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -484,7 +484,7 @@ private void ProcessSettingsFrame(FrameHeader frameHeader)
                 // Send acknowledgement
                 // Don't wait for completion, which could happen asynchronously.
-                Task ignored = SendSettingsAckAsync();
+                _ = SendSettingsAckAsync();
```
Dumb question: why is this better?

stephentoub replied:
Not a dumb question. It's the exact same resulting IL, so from a results/performance perspective it's not better.  It's just taking advantage of the C# discards feature so you don't need to name something you're not using.  Then as the discards feature gets broader adoption, it should become obvious what this is doing, if it's not already.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -106,14 +106,23 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                        {
+                            // If we still processing response after receiving response headers, this will give us chance to propagate exception up.
+                            // Since we failed while Copying stream, wrapp it as IOException if needed.
+                            _abortException = e;
```
It looks like we could end up trying to race to set _abortException.  Presumably if there are multiple exceptions, we don't care which one wins?  Or are there situations where we should be prioritizing one over another?

wfurt replied:
I don't know about any case where it would matter. Maybe @geoffkizer can comment on it but my assumption was that first exception would be closer to root cause and second would maybe be result of it. I guess we could set them atomically and nest but I'm not sure if that would be useful.

geoffkizer replied:
I'm honestly not sure how _abortException is used. Can someone explain what this is for?

wfurt replied:
That is mechanism how to propagate errors from background tasks  - like ProcessIncomingFramesAsync(). When exception happens we save it and than later on we check if something happened. In this case when response headers are already given back to caller, we use that to propagate errors from sending request body (of still pending) to response. This is streategy we discussed while back for the duplex send/receive logic.

geoffkizer replied:
Ok. So this will be something like an Http2ProtocolException or an IOException, etc? And then it becomes the inner exception for HttpRequestException somewhere?

I guess my main question here is, why are we setting this here? Seems like it should be set at a lower level, in one of the connection read or write methods.

wfurt replied:
in general, you are right.  For exceptions caused by IO or protocol, this should be already set. This was raised in context of the concurrent  body send  while receiving response headers and response body. If the body send fails for reasons like OperationCanceled or ObjectDisposed or some other reasons, this ensures we will be able to propagate that  error while processing response body. This is generally fall-back for cases when we already process response headers and return them to caller together with response stream.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -733,32 +733,62 @@ private async Task AcquireWriteLockAsync(CancellationToken cancellationToken)
-            FinishWrite(mustFlush: true);
+                FinishWrite(mustFlush: true);
+            }
+            catch (Exception e)
```
Why can't we handle this in FinishWrite, instead of always having to do try/catch in the callers? Is the exception coming from somewhere else? Seems like the only exception that should matter here (and elsewhere) is an exception from FinishWrite.

wfurt replied:
how would you exactly envision that @geoffkizer? 
Righty now I only updated few functions which are always "ignored" but in normal case the  exception  is just thrown up the call stack. Also right now we would observe exception when trying to send reset but failure to do so, would not cause exception on  it's own - if for example server is already gone.

geoffkizer replied:
Well, it seems to me we should be handling this in FinishWrite already. After all, the actual write may occur asynchronously from this code.

Let me ask the question this way: Why is this try/catch needed? You added it for some reason. Is it just to ensure that no exception escapes here? Why is that necessary?

wfurt replied:
in previous change @stephentoub raised concern about UnobservedTaskExceptions. So I look for any background task which can fail and I tried to either observe the exception (and log it) or propagate it up.

It seems like in past UnobservedTaskExceptions would kill the process. Now it won't, but we are still trying to avoid it. So I tried to make sure any exception in "ignored" task can be observed.

geoffkizer replied:
Some related observations here:

It's imperative that there's no exception between StartWriteAsync and FinishWrite in any of these routines. If that happened then the write lock would never be released.

So in theory, the only thing that can throw here is StartWriteAsync or FinishWrite. But in practice, I don't think FinishWrite actually ever throws either -- the write may happen asynchronously, so there's no real point in surfacing exceptions here. And StartWriteAsync should really only throw in the connection is already aborted.

So again, I'm not sure what we gain by having try/catch here...

wfurt replied:
ok, I'll take another look. If that is base mode of operation, maybe I can add note to make it clear.
I did not come up with any good test case for this so it was mostly precaution for cases @stephentoub was concerned about.

On that note, how failures in write would be propagated up?

wfurt replied:
The StartWriteAsync () calls  AcquireWriteLockAsync()

```c#
  if (_disposed)
  {
     throw new IOException(SR.net_http_request_aborted);
  }
```
we have seen this in unit tests and it seems like this can be way how background task gets to faulted state. In this case there is not much to do as the connection is already being aborted but I think it will lead to condition @stephentoub  was concerned about.

geoffkizer replied:
I'm still not clear why we need all these new try/catch blocks.

StartWriteAsync can throw, but if so, it will release the writer lock.

Nothing between StartWriteAsync and FinishWrite should throw.

If FinishWrite throws, it should handle whatever cleanup is necessary, e.g. call Abort().

geoffkizer replied:
Is the whole point here just to avoid UnobservedTaskExceptions? Is this really worthwhile? @stephentoub 

If it's really worth doing, then instead of modifying all these routines like SendSettingsAckAsync to never throw, we should simply add some sort of helper in the *caller* routines to eat the exception someway.

The point here is that all the Send*Async methods should have consistent behavior, and that behavior is to throw if the connection has been aborted (in StartWriteAsync) before we get a chance to grab the write lock. Some callers of these routines will care about this. Some won't. The callers themselves should decide whether they care (by awaiting) or not (by explicitly eating the exception).

geoffkizer replied:
In fact, you are already doing this in a bunch of places. You're calling IgnoreExceptionsAsync on the Task returned from SendRstStreamAsync. Why can't we just do this other places as well?

stephentoub replied:
> Is the whole point here just to avoid UnobservedTaskExceptions? Is this really worthwhile? 

Ideally.  Developers do subscribe to this event in order to find places that they've allowed exceptions to go unhandled, and we hear from them when our exceptions clutter up their results.  It's also possible (though unusual) for a developer to change a setting that causes the app to crash for such unobserved exceptions (that was the default behavior in .NET Framework 4 until we switched it to not crash in .NET Framework 4.5 when async/await was introduced.)

That said, if suppressing it adds measurable cost or complexity, it's likely not worth it, at least not until we hear about it.

geoffkizer replied:
> That said, if suppressing it adds measurable cost or complexity, it's likely not worth it, at least not until we hear about it.

I'm fine if we just add IgnoreExceptionsAsync at the calling sites. That seems to be what we do in other situations, e.g. similar cases for HTTP/1.1.

What I don't want to do is change the semantics of the actual Send*Async routines, which is what the current change does.

wfurt replied:
That will be in next update @geoffkizer. I felt that is better than pollute code with the try/catch. As you mentioned reasons for failures should be already propagate from connection IO.

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/corefx into duplex2

wfurt committed:
updates

wfurt committed:
add test for cancelation

wfurt committed:
log responseHeadersTask if bodyTask fails first

wfurt commented:
I pushed some changes to address  lingering issues raised through this conversation or related issues:

StartWriteAsync can fail as we have seen in #38481 and #35466
There is at least race (as RFC also mentioned) condition like:

- AddStream() in SendHeadersAsync()
- We get GOAWAY from server. It has no clue about new stream so we can do cleanup in CheckForShutdown()
- SendHeadersAsync() or SendRequestBody tries to continue and fails while trying to get credit.

In some of the traces, we probably got GOAWAY and failed right at RequestCreditAsync() SendHeadersAsync()

To address this I moved disposal of CreditManager objects from CheckForShutdown() to actual Dispose, so they follow life cycle of Http2Connection instead of being possibly disposed by GOAWAY frame. 

On that note, if somebody Disposes HttpClient -> Http2Connection we are still on same situation. 
If we want to avoid UnobservedTaskExceptions in that case I think added try/catch is needed. 

With #34638 tracking remaining work for GOAWAY cases I think this fixes #35466

I added missing await to collect error on responseHeadersTask
fixes #38293 

As I addresses comments from @geoffkizer I'm not aware of any remaining work for #35652
I also added test case  for #38391 HTTP2: SendAsync cancellation token does not send RST_STREAM 

In that part I bumped to #9071 where cancellation basically does not work for HttpContent.
I don't know if that is problem for gRPC but our low-level Stream function get CancelationToken.None because we use _request.Content.CopyToAsync() in SendRequestBodyAsync()

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L316-L327

For now, I simply extended test SlowTestStream() to do direct cancelation there so I can verify RST scenario. 

fixes #38391
fixes #35652

Cancelation may need some more work but that should be probably separate PR . 

It would be great if you can take another look @stephentoub and @geoffkizer 

cc: @scalablecory @davidsh @krwq @JamesNK

wfurt committed:
improve cancelation

wfurt commented:
After chatting with @stephentoub, I pushed new update to:
- cancel properly streaming RequestBody despite #9071
(did we missed 3.0 API review train on #9071 ????)
- add cancelation to ReadResponseHeadersAsync() (pointed out while back by @davidsh) as well as cancelation handling to response stream (and new test for it)

geoffkizer commented:
> To address this I moved disposal of CreditManager objects from CheckForShutdown() to actual Dispose, so they follow life cycle of Http2Connection instead of being possibly disposed by GOAWAY frame.

How does this help? If I understand correctly, you've made it so that we will actually dispose CreditManager and the stream Dictionary earlier. That seems like it makes things worse, not better.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -452,7 +466,24 @@ public async Task ReadResponseHeadersAsync()
                 if (wait)
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
+                    using (cancellationToken.Register(s => {
```
This cancellation logic seems like it should apply everywhere that GetWaiterTask is called, e.g. ReadDataAsync.

geoffkizer replied:
In other words, GetWaiterTask should take a CancellationToken, and throw if the CancellationToken is cancelled before the waiter is signalled.

There's some logic to do this with a regular TaskCompletionSource here: https://github.com/dotnet/corefx/blob/af562e84b8b5a4a335ceb5006886a6c42c1c023e/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs#L10

However, we're not using a regular TaskCompletionSource here. We're using the resettable source. I'm not quite sure the best way to make this work with cancellation.

@stephentoub thoughts?

stephentoub replied:
If we need this logic in other places, I agree with @geoffkizer it'd be good to centralize all of this into GetWaiterTask.  The downside to that is it'll make GetWaiterTask more expensive.  But maybe that doesn't matter.

geoffkizer replied:
@stephentoub Any suggestions re how to make GetWaiterTask cancellable? I'm unclear about the interactions with the resettable task stuff here.

> If we need this logic in other places, I agree with @geoffkizer it'd be good to centralize all of this into GetWaiterTask

We do -- see ReadDataAsync for example, we're ignoring the cancellationToken there currently.

> The downside to that is it'll make GetWaiterTask more expensive. But maybe that doesn't matter.

Only when a cancellationToken is specified, right? (I realize that's the default behavior with HttpClient, but at least it's possible to avoid this if it really matters)

wfurt replied:
When I talk to @stephentoub yesterday, it seems like we always pass cancellationToken token in even if user does not. (To implement timeout and cancelation of all pending requests)

krwq commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:
```diff
@@ -1466,6 +1572,67 @@ public async Task sendAsync_ConcurentSendReceive_Ok(bool shouldWait, HttpStatusC
+                         string responseBody = await response.Content.ReadAsStringAsync();
+                         Assert.True(false, "Should not be here");
+                    }
+                    catch (HttpRequestException e)
```
AFAIR Assert.Throws returns thrown exception

davidsh replied:
Both Assert.Throws<T>() and Assert.ThrowsAsync<T>() give back the exception as the return.

wfurt commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -98,22 +98,36 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                         {
-                            await _request.Content.CopyToAsync(writeStream, null, cancellationToken).ConfigureAwait(false);
+                            // TODO: until #9071 is fixed, cancellation on content.CopyToAsync does not work.
+                            // To work around it, register delegate and set _abortException as needed.
```
CopyToAsync() does take CancelationToken. But than it calls TryGetBuffer() and that fails on steaming content so we fall to 
```c# 
SerializeToStreamAsync(stream, context, cancellationToken);
```
and that is 
```c# 
internal virtual Task SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken) => SerializeToStreamAsync(stream, context);
```

so as far as I can tell we do loose provided cancelation token. 

There may be other ways to do this but there is already code check and throw if the _abortException is set. So the registration simply sets it and we will throw it later when processing the stream.

geoffkizer replied:
Can't we achieve the same thing by just checking the cancellationToken after the CopyToAsync is complete? I.e., no need to actually register here.

wfurt replied:
We could but that means we need to finish sending request body, right? For gRPC duplex streaming, that effectively means we cannot cancel request, right?

stephentoub replied:
How does this particular registration help?  When cancellation is requested it'll end up setting _abortException to an OperationCanceledException... but what's paying attention to _abortException?  Is something in the CopyToAsync operation monitoring _abortException somehow?

wfurt replied:
yes. That was part of your #37223 @stephentoub .  We will pick it up with HttpStream in read/write. If there is better way how to propagate cancelationToken itself down, let me know.

stephentoub replied:
> We will pick it up with HttpStream in read/write

But only between calls, right?  Meaning, if cancellation is requested we'll end up noticing it after a read or after a write, but if we're in the middle of the read or write, we won't actually notice it until that operation completes, correct?

Without a public overload, I don't know of a better way to achieve it, though.  The cancellation token passed into CopyToAsync below will help for some of our own content types, but most won't be given it, and thus this token won't be passed into the the constituent read and write operations.  We can address that with HTTP/1.1 by tearing down the connection on cancellation, but we can't do that here.

wfurt replied:
Yes, that is correct. As you mentioned, I did not find better easy way. So I take it as opportunistic improvement. 

Having that said, adding overload with cancelation should not be controversial so I'm wondering if there is still chance to push it to 3.0. I guess that would be primary choicer if this was on our radar, right?

stephentoub replied:
> adding overload with cancelation should not be controversial so I'm wondering if there is still chance to push it to 3.0. I guess that would be primary choicer if this was on our radar, right?

@terrajobst, @danmosemsft, could we still get something in here?

geoffkizer replied:
Okay, I understand the issue with SerializeToStreamAsync not actually being cancellable. Thanks for clarifying.

But I still don't see why we need to register here. Why not just do cancellationToken.ThrowIfCancellationRequested (or whatever) after CopyToAsync completes?

Registration is only useful if we're actually going to cancel an in-process operation, which is not the case here.

wfurt commented:
> > To address this I moved disposal of CreditManager objects from CheckForShutdown() to actual Dispose, so they follow life cycle of Http2Connection instead of being possibly disposed by GOAWAY frame.
> 
> How does this help? If I understand correctly, you've made it so that we will actually dispose CreditManager and the stream Dictionary earlier. That seems like it makes things worse, not better.

I may need to update the code little bit as I chat with @stephentoub afterwards and we want existing streams working even after Http2Connection is disposed. (which I did not plan for) The main reason to move this @geoffkizer was to not trigger it from processing GOAWAY as we may hit race condition.

I'm wondering if we can do same as you did for _writerLock and simply let it go when last reference is gone instead of disposing them explicitly. 

If we want streams to be process even after Http2Connection is disposed, can we alternatively keep CreditManager same way and keep it (partially?) functional after disposed? 

I can add some more test around disposing scenarios.

geoffkizer commented:
> I'm wondering if we can do same as you did for _writerLock and simply let it go when last reference is gone instead of disposing them explicitly.

There's an issue for figuring out how to dispose the _writerLock properly, right?

Regardless, I agree that these should be disposed at the same time the _writerLock is disposed. Since we're not doing this currently, I think it's fine to not dispose these at all for now and fix them up as part of the _writerLock fix. We should make a note in the _writerLock issue to fix these as well, and update the related comment.

wfurt committed:
feedback from review

wfurt commented:
I made some updates and added more tests  @geoffkizer .
- If we dispose HttpHandler->Http2Connection while we have pending Request or if we still have SendRequestBodyAsync() request will fail. Even if we already received response headers, we would try to propagate error to response stream.
- if are done sending request, we can finish processing response. I added test for that. 

I think problematic case was when we pick connection from pool and we receive GOAWAY or call Abort/AbortStreams before we can create new stream. In that case _disposed is set to true and CreditManager will fail. 
In past, if connection did not have any active streams, that would also dispose the writeLock.

AddStream() already had logic to deal with race and throw exception with `allowRetry: true`.
I extended that logic so we either create stream and the check for stream count should work or we will fail and retry unless we hit IO or protocol exception.

wfurt committed:
update SendAsync_ConcurentSendReceive_Fail test

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -968,12 +990,22 @@ private void WriteHeaders(HttpRequestMessage request)
+                // write lock
+                await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (ObjectDisposedException)
```
We can't avoid throwing this in the first place?

wfurt replied:
I think we can. I would suggest to do that as part #38267.

geoffkizer replied:
If we don't think this is the right change, then why are we doing anything here right now? Why is it necessary to make a temporary change we will back out later?

wfurt replied:
we have randomly failing tests and that impacts other developers. We also seen this with gRPC and we want to stabilize for Monday preview cutoff

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -968,12 +990,22 @@ private void WriteHeaders(HttpRequestMessage request)
+                // Throw a retryable request exception if this is not result of some other error.
+                // This will cause retry logic to kick in and perform another connection attempt.
+                // The user should never see this exception.  Same logic lives in AddStream.
+                throw _abortException ?? new HttpRequestException(null, null, allowRetry: true);
```
It'd be better not to do `throw _abortException`; doing that will overwrite the stack trace and watson bucket information in the exception with the current details.  If you instead do `ExceptionDispatchInfo.Throw(_abortException)` then the previous details will be retained and the current location will be appended to the exception rather than overwriting its state.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -452,7 +466,24 @@ public async Task ReadResponseHeadersAsync()
+                                signalWaiter = stream._hasWaiter;
+                                stream._hasWaiter = false;
+                            }
+                            if (signalWaiter) stream._waitSource.SetResult(false);
```
I'm not 100% sure this is safe.  It's been a while since I looked at the code for this but will want to spend some time staring at all the relevant uses tomorrow.

wfurt replied:
It would be great if you can take another  look @stephentoub. This mimics pattern used in other places where we check if somebody is waiting and set it to false under lock and that we signal waiter outside of the lock.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -452,7 +466,24 @@ public async Task ReadResponseHeadersAsync()
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw new OperationCanceledException();
+                    }
```
```C#
CancellationHelper.ThrowIfCancellationRequested(cancellationToken);
```

geoffkizer commented:
> * If we dispose HttpHandler->Http2Connection while we have pending Request or if we still have SendRequestBodyAsync() request will fail. Even if we already received response headers, we would try to propagate error to response stream.
> * if are done sending request, we can finish processing response. I added test for that.

Why is this the desired behavior? Why fail if we're still sending the request body, but allow it to continue if we're still receiving the response body?

Http2Connection.Dispose is called when the client (HttpConnectionPool) decides it no longer needs the connection. Usually this means that the connection is already dead (because the server terminated it) or idle, so in this case dispose should be simple -- just tear down the underlying socket/SslStream etc.

The somewhat more interesting case is when the HttpClient (or HttpClientHandler) is disposed. In this case, I believe the Http2Connection will be disposed immediately. However, I believe we want outstanding requests to complete in this case. This should all match the HTTP/1.1 behavior here. In fact that's how the behavior here was designed in the first place, to match the HTTP/1.1 and HttpConnection behavior with respect to Dispose. Ideally we'd have a version-agnostic test that validates this behavior for both HTTP/1.1 and HTTP/2.

There is some additional weirdness here with handling GOAWAY from the server. I think we're disposing the connection in this case and setting _disposed to true. That's probably bad, and should change. We should track whether we've received GOAWAY separately from whether the connection has been disposed by the client. See also https://github.com/dotnet/corefx/issues/38558. @krwq thoughts?

> I think problematic case was when we pick connection from pool and we receive GOAWAY or call Abort/AbortStreams before we can create new stream. In that case _disposed is set to true and CreditManager will fail.

Agreed -- see above

geoffkizer commented:
> AddStream() already had logic to deal with race and throw exception with `allowRetry: true`.
> I extended that logic so we either create stream and the check for stream count should work or we will fail and retry unless we hit IO or protocol exception.

This does seem like a problem, but it seems like a problem larger than just sending headers (which is where you added the logic). In theory, we could send all the headers and the whole request body, and as long as the server hasn't started sending the response, we can still receive a GOAWAY telling us that it didn't process the request. Not sure what to do about this, but the current change seems inadequate. Perhaps we should open a separate issue on this.

wfurt commented:
Logically, everything what Started with SendAsync(request, CT) will be subject to CT cancelation and that includes SendRequestBodyAsync() task. 
It actually may be hard no to as provided CT is handed down to the function. 
We try to mimimc HTTTP/1 behavior but there we do not have parallel send/receive so request body will be always completed before SendAsync() return.   

So I added some tests as well as enabled old test covering both HTTP/1 and HTTP/2. 
There is one more test I want to update but that will require changes to GenerecLoppback so I would prefer to do it as follow-up change.

wfurt committed:
feedback from review

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/corefx into duplex2

wfurt committed:
update SendSettingsAckAsync

wfurt commented:
There may be better way how to implement cancelation in  GetWaiterTask() but I wanted to get pending changes in and merge with loopback server change.

stephentoub commented:
I will review it all again today.

krwq commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -555,7 +556,7 @@ private void ProcessPingFrame(FrameHeader frameHeader)
             // Don't wait for completion, which could happen asynchronously.
-            Task ignored = SendPingAckAsync(_incomingBuffer.ActiveMemory.Slice(0, FrameHeader.PingLength));
+            _ = LogExceptionsAsync(SendPingAckAsync(_incomingBuffer.ActiveMemory.Slice(0, FrameHeader.PingLength)));
 
```
Irrelevant to this PR: is _incomingBuffer guarded against being modified by multiple threads at the same time? I'm wondering if we can get into some inconsistent state

geoffkizer replied:
Good catch. I do think there's an issue here. We should copy the ping data out of _incomingBuffer before we call SendPingAckAsync.

geoffkizer replied:
Filed issue #38788 to track this

krwq commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -968,12 +969,22 @@ private void WriteHeaders(HttpRequestMessage request)
+            try
+            {
+                // Ensure we don't exceed the max concurrent streams setting.
+                await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
```
(not familiar with credit manager) do credits need to be returned?

wfurt replied:
it currently "blocks" until credit is available.

wfurt committed:
more HTTP/2 cancelation tests

wfurt commented:
When I originally enabled, Dispose_DisposingHandlerCancelsActiveOperationsWithoutResponses, I did not realized it is actually not testing HTTP/2 as it was using old loopback server. 

I updated GenericLoopbackServer with four primates allowing to read request headers/body and send headers/body and I fixed test to work with both protocol versions. 

I also updated PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly @stephentoub pointed as base for cancelation testing.

wfurt committed:
update test to accept any derivative from OperationCanceledException

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/corefx into duplex2

krwq commented on src/Common/tests/System/Net/Http/GenericLoopbackServer.cs:
```diff
@@ -37,9 +38,34 @@ public abstract class GenericLoopbackServer : IDisposable
         public abstract Task<HttpRequestData> HandleRequestAsync(HttpStatusCode statusCode = HttpStatusCode.OK, IList<HttpHeaderData> headers = null, string content = null);
 
+        // Accept a new connection, and hand it to provided delegate.
+        public abstract Task AcceptConnectionAsync(Func<GenericLoopbackConnection, Task> funcAsync);
```
should it just return `Task<GenericLoopbackConnection>`?

wfurt replied:
not this one as it does not return connection but executes delegate with the connection. See http/1 in Loopback.cs

krwq commented on src/Common/tests/System/Net/Http/GenericLoopbackServer.cs:
```diff
@@ -62,6 +88,7 @@ public class HttpRequestData
         public string Method;
         public string Path;
         public List<HttpHeaderData> Headers { get; }
+        public IntPtr Handle;
```
Could you add some comment what's this Handle? (or rename it) Does not seem immediately obvious

krwq commented on src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -522,6 +530,7 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
             Assert.Equal(FrameFlags.EndHeaders, FrameFlags.EndHeaders & headersFrame.Flags);
 
             int streamId = headersFrame.StreamId;
+            requestData.Handle = (IntPtr)streamId;
```
should the handle be just named StreamId and be `int`/`int?`?

wfurt replied:
I could but purposely I did not want to make it HTTP/2 specific.

krwq replied:
fair enough, at least some comment above the handle should be sufficient

geoffkizer replied:
I'm sympathetic to the idea of this not being HTTP2 specific, but I'm not sure that making it an IntPtr actually helps anything here.

I would suggest making it an int and calling it RequestId. "Handle" is extremely vague. "RequestId" at least conveys something about the meaning intended here.

We will never use this for HTTP/1.1, so it only really matters if we need it for something like HTTP/3, and we'll need to revisit then anyway.

wfurt replied:
yes. I like that idea @geoffkizer.

krwq commented on src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -578,7 +587,7 @@ public async Task SendDefaultResponseAsync(int streamId)
         }
 
-        public async Task SendResponseHeadersAsync(int streamId, bool endStream = true, HttpStatusCode statusCode = HttpStatusCode.OK, bool isTrailingHeader = false, IList<HttpHeaderData> headers = null)
+        public async Task SendResponseHeadersAsync(int streamId, bool endStream = true, HttpStatusCode statusCode = HttpStatusCode.OK, bool isTrailingHeader = false, bool endHeaders = true, IList<HttpHeaderData> headers = null)
```
would it make sense to instead take flags directly? i.e. `FrameFlags flags = FrameFlags.EndHeaders`

krwq commented on src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -215,6 +217,12 @@ public void ShutdownSend()
         {
+            if (_shutdownDone)
+            {
+                return;
```
do we expect this to be called multiple times? In most of the cases this is presumably a test issue. Perhaps makes sense to track this in the test which needs this directly

wfurt replied:
it is also called from Dispose(). I may update Dispose() to call your new function to ignore errors but allowing it called multiple times may simplify cases with errors.

geoffkizer replied:
I agree with @krwq here -- if you're calling this twice, it's probably a test issue, so it's better not to allow it to be called twice.

wfurt replied:
I updated Dispose() to call newly added ShutdownIgnoringErrorsAsync() and removed the check.

krwq commented on src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -624,7 +633,42 @@ public async Task SendResponseBodyAsync(int streamId, ReadOnlyMemory<byte> respo
+
+        public override void Dispose()
+        {
+            WaitForConnectionShutdownAsync().GetAwaiter().GetResult();
```
is this equivalent to just `.Wait()`?

davidsh replied:
Not exactly.  It is better to use than `.Wait()` in many cases.  Using `.Wait()` will throw AggregateException with real exception inside it.  But using `.GetAwaiter().GetResult()` won't. It will only throw real exception.

krwq commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -168,6 +168,14 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
 
+        public override async Task AcceptConnectionAsync(Func<GenericLoopbackConnection, Task> funcAsync)
+        {
+            using (Http2LoopbackConnection connection = await EstablishConnectionAsync().ConfigureAwait(false))
```
won't this immediately dispose connection after accepting? (since we added Dispose it will now immediately shutdown, is that expected?)

wfurt replied:
This simply implements old behavior. Remaining it will make test cleanup more difficult. Having said that, I don't really have strong feeling about naming and I'm wrong person to argue about english.

krwq replied:
fair enough, this kind of structures are often found in functional languages and are often named with something starting with `With` to make it clear that the lifetime of the object is scoped to the body of the lambda

krwq commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -660,52 +665,129 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                {
+                    buffer = new byte[0];
+                }
+                if (_contentLength > 0)
```
nit: enter before `if`

wfurt commented:
BTW @JamesNK helped me verify that this change also fixes #38711 and we get OperationCanceledException on cancelation instead of ObjectDisposed on the stream.

davidsh commented:
/azp run

davidsh commented:
@wfurt What is the latest status on this PR?  Is it ready for merge? Some of the issues raised here can be done as a follow-up PR.  We need to get this functionality in for Preview 7.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -635,10 +651,29 @@ public void Cancel()
+                }, this))
+                {
+
+                    await vt;
```
.ConfigureAwait(false)

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -635,10 +651,29 @@ public void Cancel()
+                        signalWaiter = stream._hasWaiter;
+                        stream._hasWaiter = false;
+                    }
+                    if (signalWaiter) stream._waitSource.SetException(new OperationCanceledException());
```
The synchronization here is making me really nervous.  I don't think this makes things worse, so we shouldn't hold up this PR as a result.  But I think we need to undo the ManualResetValueTaskSource changes I'd previously made and go back to using a TaskCompletionSource for now, and revisit that post-3.0.  Unless I'm missing something, there are subtle race conditions here that could result in multiple threads accessing the MRVTS concurrently, namely one thread calling Reset on it while another calls SetResult/Exception on it, which is not supported and could lead to corruption and hangs.  Unfortunately I don't think we have sufficient testing in place to catch such issues, e.g. I don't believe we have any tests that stress using the same HttpClient instance concurrently against the same server on the same connection and with multiple access patterns.  I will try to make time this weekend after this goes in to submit a PR to at least make it safe, and then we can revisit making it fast.

stephentoub replied:
Actually, looking at it more, I think I've convinced myself it is actually safe. The saving grace (which I probably knew when I added the MRVTSC but neglected to comment) is that there can only be one consumer of the waitSource, the code consuming the response and waiting for data to show up. There would be two potential issues:
1. If multiple consumers could race to consume the waitSource
2. If multiple producers could race to set the waitSource
3. If the consumer could race to reset the waitSource concurrently with a producer completing one.

(1) can't happen because there's only ever one consumer. Parallel usage of a response stream is a bad programming error that can result in corruption, but that's the case for most data structures in .NET, and such usage is explicitly not supported (nor easy to accidentally do).

(2) is handled by the lock, which guards  extracting the waitSource and setting the Boolean hasWaiter appropriately.

(3) is the hardest to see, but is addressed again by their only being one consumer. If there could be multiple consumers, then while the producer was calling SetResult outside of the lock, a consumer could come along and Reset it inside of the lock. But since there can only be one consumer, if there is a waiter, by definition it is the sole consumer, which means a consumer couldn't concurrently be calling Reset with SetResult because it's SetResult that would wake up the sole consumer. And if there wasn't a consumer, then hasWaiter would be false, and the producer wouldn't try to SetResult.

I will subsequently add all of this to a comment in the code.

geoffkizer replied:
Additionally, once the cancellationToken is set, we should never have another waiter (past the one that's cancelled, if there is one). At this point the whole response is cancelled and so no one should be attempting to call GetWaiterTask again. I think this helps, but I'm not entirely sure...  

It might be nice to add an assert about this.

wfurt replied:
thanks @stephentoub for deeper look.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -635,10 +651,29 @@ public void Cancel()
             void IValueTaskSource.GetResult(short token) => _waitSource.GetResult(token);
+            private async Task GetWaiterTask(CancellationToken cancellationToken)
+            {
+                var vt = new ValueTask(this, _waitSource.Version).AsTask();
```
This AsTask should not be necessary.

stephentoub replied:
This wasn't addressed.  I'll fix it.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -693,6 +734,11 @@ public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationT
 
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return new ValueTask<int>(Task.FromException<int>(new OperationCanceledException()));
```
```C#
return new ValueTask<int>(Task.FromCanceled(cancellationToken));
```

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -738,6 +784,12 @@ public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationTo
+                    // TODO: until #9071 is fixed
+                    if (http2Stream._abortException is OperationCanceledException)
+                    {
+                        throw http2Stream._abortException;
```
```C#
ExceptionDispatchInfo.Throw(http2Stream._abortException);
```

stephentoub replied:
Why do we propagate this exception only if it's an OCE?  Why not any time it's not null?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionBase.cs:
```diff
@@ -24,5 +24,30 @@ public bool LifetimeExpired(long nowTicks, TimeSpan lifetime)
+        }
+
+        /// <summary>Awaits a task, logging any resulting exceptions (which are otherwise ignored).</summary>
+        internal async Task LogExceptionsAsync(Task task)
```
I suggest making this be:
```C#
internal void LogExceptions(Task task)
{
    _ = LogExceptionsAsync(task);

    async Task LogExceptionsAsync(task)
    {
        ... // current implementation
    }
}
```
That way, call sites can just do `LogExceptions(...);` rather than `_ = LogExceptionsAsync(task);`.

geoffkizer replied:
Can we do the same for IgnoreExceptionsAsync as well?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -968,12 +969,22 @@ private void WriteHeaders(HttpRequestMessage request)
+                // Throw a retryable request exception if this is not result of some other error.
+                // This will cause retry logic to kick in and perform another connection attempt.
+                // The user should never see this exception.  Same logic lives in AddStream.
+                ExceptionDispatchInfo.Throw(_abortException ?? new HttpRequestException(null, null, allowRetry: true));
```
This should instead be:
```C#
if (_abortException != null)
{
    ExceptionDispatchInfo.Throw(_abortException);
}
throw new HttpRequestException(null, null, allowRetry: true);
```

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -98,22 +98,36 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+
+                        if (_abortException == null)
+                        {
+                            // If we still processing response after receiving response headers, this will give us chance to propagate exception up.
```
Nit:
"If we still processing response" => "If we are still the response"
"this will give us chance" => "this will give us a chance"

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -542,7 +557,8 @@ public int ReadData(Span<byte> buffer)
                     Debug.Assert(bytesRead == 0);
-                    GetWaiterTask().AsTask().GetAwaiter().GetResult();
+                    GetWaiterTask(cancellationToken).GetAwaiter().GetResult();
+                    CancellationHelper.ThrowIfCancellationRequested(cancellationToken);
```
Seems like this line should be in GetWaiterTask as the last thing it does?

wfurt replied:
done

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -98,22 +98,36 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
-                            await _request.Content.CopyToAsync(writeStream, null, cancellationToken).ConfigureAwait(false);
+                            // TODO: until #9071 is fixed, cancellation on content.CopyToAsync does not work.
+                            // To work around it, register delegate and set _abortException as needed.
+                            using (cancellationToken.Register(stream => { if (((Http2Stream)stream)._abortException == null) ((Http2Stream)stream)._abortException = new OperationCanceledException(); }, this))
```
This can use UnsafeRegister instead of Register. Nothing in the callback needs ExecutionContext flowed.

stephentoub replied:
You marked this as resolved, but it wasn't addressed.

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -624,7 +633,42 @@ public async Task SendResponseBodyAsync(int streamId, ReadOnlyMemory<byte> respo
+        {
+            int streamId = handle == (IntPtr)0 ? _lastStreamId : (int)handle;
+            bool endHeaders = body != null || isFinal;
+            return SendResponseHeadersAsync(streamId, endStream: isFinal, statusCode, isTrailingHeader : false, endHeaders : endHeaders, headers);
```
No space before the ':' in arg lists

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1227,14 +1238,6 @@ private void CheckForShutdown()
             _connectionWindow.Dispose();
             _concurrentStreams.Dispose();
-
-            // ISSUE #35466
```
Why was this comment removed? We haven't actually fixed the issue here, have we?

wfurt replied:
I think the flow is understood and remaining work tracked by other existing issues. If you have specific 3.0 task let me know or we can keep it open for future investigation.

wfurt committed:
feedback from review

wfurt commented:
/azp run

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/corefx into duplex2

wfurt commented:
/azp run

wfurt committed:
roll-back changes in AcceptConnectionAsync

wfurt commented:
/azp run

wfurt commented:
Windows.Debug failed in System.Security.Cryptography.Pkcs
